hub/api/dataset.py

```diff
@@ -101,6 +110,20 @@ def __iter__(self):
         for i in range(len(self)):
             yield self[i]
 
+    def flush(self):
+        """Necessary operation after writes if a cache is being used. Clears the cache and sends the data to underlying storage"""
```

Comment:
can we have this flush function called implicitly? i as a user don't want to know of `ds.flush()`'s existence

Reply:
with a cache, the only other way is to use the "with" syntax. We can discuss this in detail if you want, but I don't think there's a way around using either one of these.

Reply:
if no, what exactly is the problem you foresee? i know that if you call a write op like 1,000 times it is kind of inefficient, but this cost isn't terrible, and i think it is justified given it's more simple to use.

Reply:
The cost is literally very terrible.
Do you remember writing to s3 without cache? It'll be the same as that.

Reply:
Could we have some meet-in-the-middle so that it's not such a heavy operation? For instance:
- Flush completed chunks as they are filled
- Flush on array-writes (`ds.image = arr`) but not appends
- A switch to enable auto-flush on any update

I recognize that not having manual-flush is really bad, but it's only bad when we're doing a lot of sequential writes. If we can control things from the dataset side (as opposed to the storage provider side) we might be able to strike a balance. 
Also pinging @davidbuniat @imshashank for visibility